### PR TITLE
Adds the SAPI handler plugin definition and manager

### PR DESCRIPTION
**Pull request by laurisigaunis**
Custom plugin type to be place where work further.

Trello card - https://trello.com/c/tU87JXAg/1-handler-plugin-type-annotatoion

### CODE REVIEW

File: src/Annotation/SapiHandlerAnnotation.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
+<?php
+
+/**
+ * @file
```
Comment (maijs) [inline src/Annotation/SapiHandlerAnnotation.php]: @file docblock should not be placed in files that contain namespaced classes (see https://www.drupal.org/node/2304909 and https://www.drupal.org/coding-standards/docs#file)

File: src/Annotation/SapiHandlerAnnotation.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Defines a Sapi handler annotation item annotation object.
+ *
+ * @see \Drupal\sapi\Plugin\SapiHandlerAnnotationManager
+ * @see plugin_api
+ *
+ * @Annotation
+ */
+class SapiHandlerAnnotation extends Plugin {
```
Comment (maijs) [inline src/Annotation/SapiHandlerAnnotation.php]: I'd suggest to rename the class to `StatisticsPlugin` for now. The reason is that, while this class is a definition for annotation, it also defines how we define our plugins. IMHO `@StatisticsPlugin()` in plugin definition looks clearer than `@ SapiHandlerAnnotation()`. Update: `SapiPlugin` sounds as good.

File: src/Plugin/SapiHandlerAnnotationBase.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ */
+
+namespace Drupal\sapi\Plugin;
+
+use Drupal\Component\Plugin\PluginBase;
+
+/**
+ * Base class for Sapi handler annotation plugins.
+ */
+abstract class SapiHandlerAnnotationBase extends PluginBase implements SapiHandlerAnnotationInterface {
```
Comment (maijs) [inline src/Plugin/SapiHandlerAnnotationBase.php]: Similarly to the comment above, I'd suggest to go with `StatisticsPluginBase`. Update: `SapiPluginBase` sounds as good.

File: src/Plugin/SapiHandlerAnnotationInterface.php [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ */
+
+namespace Drupal\sapi\Plugin;
+
+use Drupal\Component\Plugin\PluginInspectionInterface;
+
+/**
+ * Defines an interface for Sapi handler annotation plugins.
+ */
+interface SapiHandlerAnnotationInterface extends PluginInspectionInterface {
```
Comment (maijs) [inline src/Plugin/SapiHandlerAnnotationInterface.php]: Similarly to the comments above, I'd suggest to go with `StatisticsPluginInterface`. Update: `SapiPluginInterface` sounds as good.

File: src/Plugin/SapiHandlerAnnotationManager.php [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+namespace Drupal\sapi\Plugin;
+
+use Drupal\Core\Plugin\DefaultPluginManager;
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Extension\ModuleHandlerInterface;
+
+/**
+ * Provides the Sapi handler annotation plugin manager.
+ */
+class SapiHandlerAnnotationManager extends DefaultPluginManager {
```
Comment (maijs) [inline src/Plugin/SapiHandlerAnnotationManager.php]: Assuming that our plugins are called (annotated as) `StatisticsPlugin`, `StatisticsPluginManager` would make a fine, fine name. Update: `SapiPluginManager` sounds as good.

File: sapi.services.yml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+services:
+  plugin.manager.statisticsplugin.processor:
```
Comment (maijs) [inline sapi.services.yml]: `plugin.manager.statisticsplugin` would make it lean and clean.

File: src/Plugin/StatisticsPluginManager.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+   *   keyed by the corresponding namespace to look for plugin implementations.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache_backend
+   *   Cache backend instance to use.
+   * @param \Drupal\Core\Extension\ModuleHandlerInterface $module_handler
+   *   The module handler to invoke the alter hook with.
+   */
+  public function __construct(\Traversable $namespaces, CacheBackendInterface $cache_backend, ModuleHandlerInterface $module_handler) {
+    parent::__construct('Plugin/StatisticsPlugin', $namespaces, $module_handler, 'Drupal\sapi\Plugin\StatisticsPluginInterface', 'Drupal\sapi\Annotation\StatisticsPlugin');
+
+    $this->alterInfo('sapi_StatisticsPlugin_info');
```
Comment (maijs) [inline src/Plugin/StatisticsPluginManager.php]: `$this->alterInfo('statistics_plugin_info');`?

Update: @laurisigaunis pointed that `alterInfo()` should have a key that starts with the module name, and that's a valid point. Given that we don't know what kind of plugin types we might come up with, it's good to be specific and reference the plugin type by `statistics_plugin`.

```
$this->alterInfo('sapi_statistics_plugins_info');
$this->setCacheBackend($cache_backend, 'sapi_statistics_plugins');
```

File: sapi.services.yml (hunk 6 above)
Comment (james-nesbitt) [inline sapi.services.yml]: this service doesn't include the name of the module that provides it.  Is that acceptable?

**Comment by james-nesbitt** [comment]
@laurisigaunis sorry that I merge other work before this.  There is now a merge conflict here (most likely services.yml.)  If you would like advice on resolving this, feel free to ask @maijs or myself.

**Comment by laurisigaunis** [comment]
PR was broken due git rebase

**Pull request closed without merging**